%% LyX 2.0.3 created this file.  For more info, see http://www.lyx.org/.
%% Do not edit unless you really know what you are doing.
\documentclass{report}
\usepackage[latin9]{inputenc}
\usepackage{listings}
\usepackage[a4paper]{geometry}
\geometry{verbose}
\setcounter{secnumdepth}{4}
\setcounter{tocdepth}{3}
\usepackage{textcomp}
\usepackage{bm}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{graphicx}
\usepackage[authoryear]{natbib}

\makeatletter

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% LyX specific LaTeX commands.
%% Because html converters don't know tabularnewline
\providecommand{\tabularnewline}{\\}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% User specified LaTeX commands.

\usepackage{array}\usepackage{bm}\usepackage{soul}\usepackage{color}\usepackage[pagewise]{lineno}\usepackage{setspace}\usepackage{listings}\usepackage{multirow}

\geometry{ hmargin=2.5cm, vmargin=2.5cm }

% \usepackage[latin1]{inputenc}
 %\usepackage[T1]{fontenc}
 %\usepackage[normalem]{ulem}
% \usepackage[french]{babel}
% \usepackage{verbatim}
% \usepackage{graphicx}

\title{\Huge \emph{\textbf{DIYABC}} \\
version 2.1\\
------\\
A user-friendly software \\
for inferring population history  through \\
Approximate Bayesian Computations\\
using\\
microsatellite, DNA sequence and SNP data}
\vspace{1.0 cm}
\author{
\Large J.M. Cornuet, P. Pudlo,\\
\Large J. Veyssier, Etienne Loire, Filipe Santos,\\
\Large A. Dehne-Garcia and A. Estoup\\
 \\
\vspace{6.0cm}
\\
Centre de Biologie et de Gestion des Populations\\
Institut National de la Recherche Agronomique\\
755 avenue du Campus Agropolis,  CS 30016\\
34988 Montferrier-sur-Lez cedex, France\\
\texttt{(diyabc@supagro.inra.fr)}}


\date{\today}
\normalsize

% Ce qui suit sert à numéroter les section, sous-sections, ... a partir de 1 au lieu de 0


\renewcommand{\thesection}{\arabic{section}.}
\renewcommand{\thesubsection}{\thesection \arabic{subsection}}
\renewcommand{\theparagraph}{\thesubsection.\arabic{paragraph}}

\let\sectionv\section
\renewcommand{\section}[1]{\sectionv{#1} \setcounter{paragraph}{0}}

\let\subsectionv\subsection
\renewcommand{\subsection}[1]{\subsectionv{#1} \setcounter{paragraph}{0}}
% Fin de la modif de renumerotation

\linenumbers


\makeatother

\begin{document}
\maketitle %\clearpage
\pagestyle{myheadings}
\markboth{DIYABC v2.1}{DIYABC v2.1}
\begin{doublespacing}
\setcounter{tocdepth}{4}
\tableofcontents{}
\end{doublespacing}
\newpage


\section{Preface}


\subsection{General context and history of DIYABC}

In less than 10 years, Approximate Bayesian Computations (ABC) have
developed in the Population Genetics community as a new tool for inference
on the past history of populations and species. Compared to other
approaches based on the computation of the likelihood which are still
restrained to a very narrow range of evolutionary scenarios and mutation
models, the ABC approach has demonstrated its ability to stick to
biological situations that are much more complex and hence realistic.
However, this approach still requires numerous computations to be
performed so that it has been used mostly by specialists (i.e. statisticians
and programmers). This has almost certainly restrained the possible
impact of ABC in population genetic studies. We believe that this
situation must be improved and therefore we have developed a computer
program for the large community of experimental biologists. We therefore
designed DIYABC as a user-friendly program allowing non specialist
biologists to achieve their own analysis.\
 \textbf{The first version (DIYABC v$0.x$)} had been written especially
for microsatellite data. There were at least two reasons for that.
The first one is that we have been among the first to develop and
use this class of markers in population genetic studies \citep[e.g.][]{E1993}.
Since then, we have developed microsatellites in numerous species
as well as we have published theoretical studies and reviews on these
markers \citep[e.g.][]{E2002}. The second reason is that microsatellites
have been and still are very popular markers in the population geneticist
community and there is now a large quantity of data that might benefit
of an ABC approach.\
 \textbf{The second version of our software (DIYABC v$1.x$)} has
been designed to make use of DNA sequence data. This has several immediate
consequences. For instance, the standard Genepop data file format
has been extended to incorporate sequence data. This has been done
in collaboration with the authors of $Genepop$ and explained in subsection
4.1.1. In this version, sequence loci are considered in the same way
as microsatellite loci, i.e. they are considered as genetically independent
and intra-locus recombination is not (yet) available. Regarding mutation
models for DNA sequences, we used the same philosophy as for microsatellites,
i.e. the program considers only simple and widely used models, keeping
in mind that a higher-dimensional parameter space will be less well
explored than a lower-dimensional space. Note that none of these mutation
models includes insertion-deletions. Also five categories of loci
(either microsatellites or DNA sequences) were considered in this
second version : autosomal diploid, autosomal haploid, X-linked, Y-linked
and mitochondrial. Note that X-linked loci can be used for an haplo-diploid
species in which both sexes have been sampled. If non-autosomal loci
have been typed in population samples, the sex-ratio of the species
will have to be provided (see subsection 4.1.1).\\


%Also for DNA sequence loci, one can choose among two possibilities : either each locus has its own mutation model or they all share the same mutation model (but not the same mutation parameters). Going back to microsatellites, there is an additional mutation parameter added to this version : the mean rate of single nucleotide insertion-deletion mutations.\\


%Given that mitochondrial DNA sequences are commonly used by population geneticists and that different effective population sizes correspond to such monoparental loci, a specific code has also been added to the genepop format to comply with this new requirement (see 4.1.1 for details).


Other improvements over version 0 included :
\begin{enumerate}
\item the use of multithread technology in order to exploit multicore/multiprocessor
computers. This is especially useful when building the reference table
and for several other intensive computation steps, such as the multinomial
logistic regression,
\item a new option which helps the detection of "bad"
prior modelisation of the data,
\item another new option which helps evaluate the goodness of fit of a given
model-parameter posterior combination (i.e. Model checking),
\item many new screens implemented not only to treat sequence data, but
also to cope with the new options described above, as well as to offer
useful complementary information on the current run.
\end{enumerate}
\textbf{The third version of DIYABC (DIYABC v$2.x$)} has been entirely
recoded in order to be used under the usual three OS (Linux, Windows
and Mac). Also the code for computations has been separated from that
of the graphic user interface (GUI). The former has been rewritten
in C++ and the latter is a mixture of Python and Qt (PyQt). The user
can then launch computations with or without using the GUI. The GUI
's uses are :
\begin{enumerate}
\item the management of projects
\item the input of the historical and genetical models
\item the parameterization of analyses
\item the launch of computations of the reference table and of the various
required analyses
\item the visualization of results
\end{enumerate}
Also, as DNA sequences have been added in the second version, a new
category of markers has been added to the third version : Single Nucleotide
Polymorphisms (SNPs). Instead of extending once more the Genepop format,
a new data simple format has been designed for these markers. Note
that SNP data are treated separately from other markers ($i.e.$ they
cannot be analyzed together with microsatellite and/or DNA sequence
data). It is worth noting that, in the present version of the program,
the analysed SNP data are assumed to correspond to independent selectively-neutral
loci, without any ascertainment bias (i.e. the deviations from expected
theoretical results due to the SNP discovery process in which a small
number of individuals from selected populations are used as discovery
panel). \\


This version includes all improvements of version 1.x and a few new
improvements such as :
\begin{itemize}
\item loci of the same type (i.e. microsatellites on one hand or DNA sequences
on the other hand) can be associated in one or more groups. This allows
for instance to define different mutation models for microsatellites
with motifs of different lengths.
\item the model checking option is now presented as a direct option (not
a suboption of the ABC estimation of parameters) which largely simplifies
its use.
\item the logistic regression can be performed on linear discriminant analysis
components instead of all summary statistics. This reduces the number
of dependent variables, thus allowing to run large "confidence
in scenario choice" analyses including many summary
statistics and scenarios \citep{EL2012}.
\end{itemize}
\textbf{\textit{\emph{The latest version of the program}}}\textbf{\textit{
(}}\textbf{\textit{\emph{DIYABC v2.1.0) includes the following major
improvements}}}: (i) new analysis options to compute error / accuracy
indicators conditionally to the observed dataset, (ii) possibility
to specify a MAF (minimum allele frequency) criterion on the analyzed
SNP datasets, and (iii) optimization of the simulation process of
SNP datasets that include a substantial amount of missing data.

\emph{1. New analysis options to compute error / accuracy indicators
conditionally to the observed dataset.}

The program DIYABC allows evaluating the confidence in scenario choice
and the accuracy of parameter estimation under a given scenario using
simulated pseudo-observed datasets (pods), for which the true scenario
ID and parameter values are known. So far such pods were drawn randomly
into prior distributions for both the scenario ID and the parameter
values. By doing so, we estimate global error/accuracy levels computed
over the whole (and usually huge) data space defined by the prior
distributions. These indicators hence actually correspond to \textquotedblleft{}prior\textquotedblright{}
error rates (when evaluating the confidence in scenario choice) or
\textquotedblleft{}prior\textquotedblright{} precision measures (when
evaluating the accuracy of parameter estimation under a given scenario).
The levels of error/accuracy may be substantially different depending
on the location of an observed or pseudo-observed dataset in the prior
data space. Indeed, some peculiar combination of parameter values
may correspond to situations of strong (weak) discrimination among
the compared scenarios or of accurate (inaccurate) estimation of parameter
values under a given model. Aside from their use to select the best
classifier and set of summary statistics, prior-based indicators are,
however, poorly relevant since, for a given dataset, the only point
of importance in the data space is the observed dataset itself. Computing
error / accuracy indicators conditionally to the observed dataset
(i.e. focusing around the observed dataset by using the posterior
distributions) is hence clearly more relevant than blindly computing
indicators over the whole prior data space as done so far. This is
basically what DIYABC v2.1.0 proposes to do with several new analysis
sub-options available within the options \textquotedblleft{}Evaluate
confidence in the scenario choice\textquotedblright{} and \textquotedblleft{}Compute
bias and precision on parameter estimations\textquotedblright{}. Indeed,
one can now choose to compute a \textquotedblleft{}posterior\textquotedblright{}
error rate (when evaluating the confidence in scenario choice) by
drawing the scenario ID and parameter values of a large number of
pods from the \textit{s} simulated datasets closest to the observed
dataset (i.e. the s datasets with the smallest Euclidean distance).
Typically, \textit{s} = 500 (when simulating 10,000 to 1 million datasets
per compared scenario) but this number can be lowered to 100. In the
same vein, one can now choose to compute \textquotedblleft{}posterior\textquotedblright{}
accuracy indicators (when evaluating the accuracy of parameter estimation
under a given scenario) by drawing the parameter values of a large
number of pods among the parameter posterior distributions estimated
under a given scenario using a standard ABC procedure. Note that we
found, using controlled genetics experiments, that posterior error
(accuracy) measures could strongly differ from prior error (accuracy)
measures, hence making a case of the significance of computing error
(accuracy) measures conditionally to the observed dataset rather than
blindly computing such measures over the whole prior data space (unpublished
results and see Pudlo et al. 2015).

\emph{2. Possibility to specify a MAF (minimum allele frequency) criterion
on the analyzed SNP datasets.}

Compared to other types of molecular markers, SNP loci have low mutation
rates, so that polymorphism at such loci results from a single mutation
during the whole population(s) gene tree and genotypes are bi-allelic.
To generate a simulated polymorphic dataset at a given SNP locus,
we proceeded following the algorithm proposed by Hudson (2002) (cf
\textendash{}s 1 option in the program ms associated to Hudson, 2002).
Briefly, the genealogy at a given locus of all genes sampled in all
populations of the studied dataset is simulated until the most recent
common ancestor according to coalescence theory. Then a single mutation
event is put at random on one branch of the genealogy (the branch
being chosen with a probability proportional to its length relatively
to the total gene tree length). This algorithm provides the simulation
efficiency and speed necessary in the context of ABC, where large
numbers of simulated datasets including numerous SNP loci have to
be generated (Cornuet et al. 2014). Most importantly, using the Hudson\textquoteright{}s
simulation algorithm is equivalent to applying a default MAF (minimum
allele frequency) criterion on the simulated dataset. As a matter
of fact, each locus in both the observed and simulated datasets will
be characterized by the presence of at least a single copy of a variant
over all genes sampled from all studied populations (i.e. pooling
all genes genotyped at the locus). In DIYABC v2.1.0, it is possible
to impose a different MAF criterion for each locus on the observed
and simulated datasets. This MAF is computed pooling all genes genotyped
over all studied population samples. For instance, the specification
of a MAF equal to 5\% will automatically select a subset of m loci
characterized by a minimum allele frequency $>$ 5\% among the \textit{l}
locus of the observed dataset. In agreement with this, only \textit{m}
locus with a MAF$>$5\% will be retained in a simulated dataset (simulated
loci with a MAF\ensuremath{\le}5\% will be discarded). In practice,
the instruction for a given MAF has to be indicated directly in the
headline of the observed dataset. For instance, if one wants to consider
only loci with a MAF equal to 5\% one will write $<$MAF=0.05$>$ 
in the headline. Writing \textsf{$<$MAF=hudson$>$ }(or omitting to write
any instruction with respect to the MAF) will bring the program to
use the standard Hudson\textquoteright{}s algorithm without further
selection as done so far in the previous version of DIYABC. The selection
with DIYABC v2.1.0 of a subset of loci fitting a given MAF allows:
(i) to remove the loci with very low level of polymorphism from the
dataset and hence increase the mean level of genetic variation of
both the observed and simulated datasets, without producing any bias
in the analyses; and (ii) to reduce the proportion of loci for which
the observed variation may corresponds to sequencing errors. In practice
MAF values \ensuremath{\le}10\% are considered. To check for the consistency/robustness
of the ABC results obtained, it may be useful to treat a SNP dataset
considering different MAFs (for instance MAF=hudson, MAF=0.01 and MAF=0.05).

\emph{3. Optimization of the simulation process of SNP datasets that
include a substantial amount of missing data.}

We have radically changed our way to take into account missing data
for SNP datasets (i.e. missing genotypes denoted \textquotedblleft{}9\textquotedblright{}
in the data file). The initial way to deal with missing data turned
out to be poorly efficient in term of computation time, especially
when the number of SNP missing data was large which seems to be the
case for many real SNP datasets. The new code we have implemented
to deal with this issue is particularly efficient and makes it feasible
to simulate in a reasonable time large SNP datasets including (or
not) numerous missing data.

%\item ascertainment bias in the design of SNPs can be tentatively corrected by considering ``reference'' samples in which the loci need to be polymorphic in order to belong to the SNP set. These reference samples are not necessarily included of the actual samples.


Finally, as for DIYABC v1, the most recent versions of DIYABC v2 (v2.0
and v2.1) deals with sexually reproducing diploid or haploid species
(co-dominant markers corresponding to autosomal, X-linked, Y-linked
loci) but does not allow considering species reproducing clonally.

For all versions of DIYABC, we recommend non-expert users to use the
GUI for their computations.


\subsection{References to cite}
\begin{itemize}
\item \textbf{version 0 :} Cornuet J.M., F. Santos, M.A. Beaumont, C.P.
Robert, J.M. Marin, D.J. Balding, T. Guillemaud and A. Estoup. 2008.
Inferring population history with DIYABC: a user-friendly approach
to Approximate Bayesian Computations. \emph{Bioinformatics}, \textbf{24}
(23), 2713-2719.
\item \textbf{version 1 :} Cornuet J.M., V. Ravign� and A. Estoup, 2010.
Inference on population history and model checking using DNA sequence
and microsatellite data with the sofware DIYABC (v1.0) (2010) \emph{BMC
Bioinformatics}, \textbf{11}, 401.
\item \textbf{version 2 :} Cornuet, J-M., Pudlo, P., Veyssier, J., Dehne-Garcia
A., Gautier M., Leblois R., Marin J-M, and A. Estoup, 2014. DIYABC
v2.0: a software to make approximate Bayesian computation inferences
about population history using single nucleotide polymorphism, DNA
sequence and microsatellite data. \textit{Bioinformatics}. Vol. 30,
no. 8, p1187\textendash{}1189, doi: 10.1093/bioinformatics/btt763.
\end{itemize}

\subsection{Web site}

http://www1.montpellier.inra.fr/CBGP/DIYABC\\
 You can get there executable files for different operating systems
as well as the last version of this detailed notice document.


\subsection{System requirements}
\begin{itemize}
\item DIYABC should run on any linux flavour, Microsoft Windows XP and Seven
and OS x 10.5 (intel) or later.
\item Minimum 4GB of RAM; 6GB of RAM recommended
\item 70MB free disk space for DIYABC binaries
\item From 1 to 10GB free disk space for each project depending on the project
configuration and the records number in the reftable file.
\end{itemize}
Caveat : it is possible that on windows 32bits (and sometimes on windows
64bits) the reference table file will not grow more than 4Go. We hope
to be able to circumvent this constraint soon on windows 64bits.


\subsection{How to create (and send) a bug report}

DIYABC V2 provides an easy way to send to the program developers the
different files and clues that are necessary to attempt solving a
bug. Click on the “Help” menu. Then go on the “Create bug report”
tab. You then just need to following the few instructions we give
you at this stage, validate and save the created bug report tarred
file somewhere on your computer. Please send the bug report file to
the indicated email address (\texttt{DIYABC@supagro.inra.fr}). Two
remarks here: (i) the bug you describe has to be the last things that
you did with the program; and (ii) please try to reproduce your bug
one time and then create and send the bug report. Finally, it had
to be noted that if the bug completely crash the application then
no bug report can be created. We will do our best to solve your bug
thanks to the bug report you provided us.


\subsection{Acknowledgements}

We thank Mark Beaumont who has been at the origin of our interest
for ABC. He offered us constant help and inspiration since the beginning.
We also thank David Balding who welcomed one of us (JMC) in his team
during the whole writing of the program and who organized several
workshops on ABC during the same period. We are indebted to Christian
Robert, Jean-Michel Marin, Stuart Baird, Thomas Guillemaud, Renaud
Vitalis, Gael Kergoat, Gilles Guillot and David Welsh with whom we
discussed many theoretical and practical aspect of DIYABC in the numerous
meetings financed by a grant from the French Research National Agency
(project $MISGEPOP$ ANR-05-BLAN-196). The same grant is also aknowledged
for having paid for the 2-year salary of FS. This research was also
supported by an EU grant awarded to JMC as an EIF Marie-Curie fellowship
(project $StatInfPopGen$) and which allowed him to come to David
Balding's place at Imperial College (London, UK). Current and future
developments of DIYABC are financed by a new grant from the French
Research National Agency (project $EMILE$ ANR-09-BLAN-0145) awarded
in september 2009. We thank several “beta-users”, especially Eric
Lombaert, Michael Fontaine, Christophe Plantamp, Marie-Pierre Chapuis,
Carine Brouat, Raphaël Leblois and Thomas Guillemaud, who tested
the the DIYABC V2 software with their data.

\newpage{}\input{methodology.tex} \input{gui_description.tex} \input{Implementation.tex}
\input{biblio.tex}
\end{document}
