\documentclass [a4paper]{report}
\usepackage{graphicx,array,natbib,amsmath,bm}
\usepackage{geometry}
\usepackage{soul,color}
\usepackage[pagewise]{lineno}
\usepackage{setspace}
\usepackage{listings}
\usepackage{multirow}

\geometry{ hmargin=2.5cm, vmargin=2.5cm }

% \usepackage[latin1]{inputenc}
 %\usepackage[T1]{fontenc}
 %\usepackage[normalem]{ulem}
% \usepackage[french]{babel}
% \usepackage{verbatim}
% \usepackage{graphicx}

\title{\Huge \emph{\textbf{DIYABC}} \\ version 2.0\\------\\ A user-friendly software \\  for inferring population history  through \\ Approximate Bayesian Computations}

\author{\vspace{1.0 cm}\Large J.M. Cornuet, P. Pudlo, J. Veyssier,\\ A. Dehne-Garcia, A. Estoup and V. Ravign\'e\\
\vspace{6.0cm}
Centre de Biologie et de Gestion des Populations\\
Institut National de la Recherche Agronomique\\
Campus International de Baillarguet, CS 30016 Montferrier-sur-Lez\\
34988 Saint-G\'ely-du-Fesc Cedex, France\\
\texttt{(diyabc@cbgp.supagro.fr)}}


\date{\today}
\normalsize

% Ce qui suit sert à numéroter les section, sous-sections, ... � partir de 1 au lieu de 0
\setcounter{secnumdepth}{4} 

\renewcommand{\thesection}{\arabic{section}.} 
\renewcommand{\thesubsection}{\thesection \arabic{subsection}} 
\renewcommand{\theparagraph}{\thesubsection.\arabic{paragraph}} 

\let\sectionv\section 
\renewcommand{\section}[1]{\sectionv{#1} \setcounter{paragraph}{0}} 

\let\subsectionv\subsection 
\renewcommand{\subsection}[1]{\subsectionv{#1} \setcounter{paragraph}{0}} 
% Fin de la modif de renum�rotation

\linenumbers
\begin{document}
\maketitle %\clearpage
\pagestyle{myheadings}
\markboth{DIYABC v2.0}{DIYABC v2.0}
\begin{doublespacing}
\tableofcontents
\end{doublespacing}
\newpage
\section{Preface}
In less than 10 years, Approximate Bayesian Computations (ABC) have developed in the Population Genetics community as a new tool for inference on the past history of populations and species. Compared to other approaches based on the computation of the likelihood which are still restrained to a very narrow range of evolutionary scenarios and mutation models, the ABC approach has demonstrated its ability to stick to biological situations that are much more complex and hence realistic. However, this approach still requires numerous computations to be performed so that it has been used mostly by specialists (i.e. statisticians and programmers). This has almost certainly restrained the possible impact of ABC in population genetic studies. We believe that this situation must be improved and therefore we have developed a computer program for the large community of experimental biologists. We therefore designed DIYABC as a user-friendly program allowing non specialist biologists to achieve their own analysis.\

The first version ($DIYABC v0.x$) had been written especially for microsatellite data. There were at least two reasons for that. The first one is that we have been among the first to develop and use this class of markers in population genetic studies \citep[e.g.][]{E1993}. Since then, we have developed microsatellites in numerous species as well as we have published theoretical studies and reviews on these markers \citep[e.g.][]{E2002}. The second reason is that microsatellites have been and still are very popular markers in the population geneticist community and there is now a large quantity of data that might benefit of an ABC approach.\

The second version of our software ($DIYABC v1.x$) has been designed to make use of DNA sequence data.This has several immediate consequences. For instance, the standard Genepop data file format has been extended to incorporate sequence data. This has been done in collaboration with the authors of $Genepop$ and explained in  subsection 4.1.1. In this version, sequence loci are considered in the same way as microsatellite loci, i.e. they are considered as genetically independent and intra-locus recombination is not (yet) available. Concerning mutation models for DNA sequences, we used the same philosophy as for microsatellites, i.e. the program considers only simple and widely used models, keeping in mind that a higher-dimensional parameter space will be less well explored than a lower-dimensional space. Note that none of these mutation models includes insertion-deletions. 
Also five categories of loci (either microsatellites or DNA sequences) were considered in this second version : autosomal diploid, autosomal haploid, X-linked, Y-linked and mitochondrial. Note that X-linked loci can be used for an haplo-diploid species in which both sexes have been sampled. If non-autosomal loci have been typed in population samples, the sex-ratio of the species will have to be provided (see subsection 4.1.1).\\ 

%Also for DNA sequence loci, one can choose among two possibilities : either each locus has its own mutation model or they all share the same mutation model (but not the same mutation parameters). Going back to microsatellites, there is an additional mutation parameter added to this version : the mean rate of single nucleotide insertion-deletion mutations.\\

%Given that mitochondrial DNA sequences are commonly used by population geneticists and that different effective population sizes correspond to such monoparental loci, a specific code has also been added to the genepop format to comply with this new requirement (see 4.1.1 for details). 

Other improvements over version 0 included :
\begin{enumerate}
\item the use of multithread technology in order to exploit multicore/multiprocessor computers. This is especially useful when building the reference table and for several other intensive computation steps, such as the multinomial logistic regression,
\item a new option which helps the detection of "bad" prior modelisation of the data, 
\item another new option which helps evaluate the goodness of fit of a given model-parameter posterior combination (i.e. Model checking),   
\item many new screens implemented not only to treat sequence data, but also to cope with the new options described above, as well as to offer useful complementary information on the current run. 

\end{enumerate}

 The third version of $DIYABC$ ($DIYABCv2.x$)  has been entirely recoded in order to be used under the usual three OS (Windows, Mac and Linux). Also the code for computations has been separated from that of the graphic user interface (GUI). The former has been rewritten in C++ and the latter is a mixture of Python and Qt (PyQt). The user can then launch computations with or without using the GUI. The GUI 's uses are : 

\begin{enumerate}
\item the management of projects
\item the input of the historical and genetical models
\item the parameterization of analyses
\item the launch of computations  of the reference table and of the various required analyses
\item the visualization of results
\end{enumerate}

Also, as DNA sequences have been added in the second version, a new category of markers has been added to the third version : Single Nucleotide Polymorphisms (SNPs). Instead of extending once more the Genepop format, a new data simple format has been designed for these markers.  

This version includes all improvements of version 1.x and a few new improvements such as :
\begin{itemize}
\item loci of the same type (i.e. microsatellites on one hand or DNA sequences on the other hand) can be associated in one or more groups. This allows for instance to define different mutation models for microsatellites with motifs of different lengths.
\item  the model checking option is now presented as a direct option (not a suboption of the ABC estimation of parameters) which largely simplifies its use.
\item the logistic regression can be performed on factorial discriminant analysis components instead of all summary statistics. This reduces the number of dependent variables, thus allowing to run large "confidence in scenario choice" analyses including many summary statistics and scenarios.
\item ascertainment bias in the design of SNPs can be tentatively corrected by considering ``reference'' samples in which the loci need to be polymorphic in order to belong to the SNP set. These reference samples are not necessarily included of the actual samples.
\end{itemize}

\subsection{Acknowledgements}
We thank Mark Beaumont who has been at the origin of our interest for ABC. He offered us constant help and inspiration since the beginning. We also thank David Balding who welcomed one of us (JMC) in his team during the whole writing of the program and who organized several workshops on ABC during the same period. We are indebted to Christian Robert, Jean-Michel Marin, Stuart Baird, Thomas Guillemaud, Renaud Vitalis, Gael Kergoat, Gilles Guillot and David Welsh with whom we discussed many theoretical and practical aspect of DIYABC in the numerous meetings financed by a grant from the French Research National Agency (project $MISGEPOP$ ANR-05-BLAN-196). The same grant is also aknowledged for having paid for the 2-year salary of FS. This research was also supported by an EU grant awarded to JMC as an EIF Marie-Curie fellowship (project $StatInfPopGen$) and which allowed him to come to David Balding's place at Imperial College (London, UK). Current and future developments of DIYABC are financed by a new grant from the French Research National Agency (project $EMILE$ ANR-09-BLAN-0145) awarded in september 2009.
\subsection{References to cite}
\begin{itemize}
\item \textbf{version 0 :} Cornuet J.M., F. Santos, M.A. Beaumont, C.P. Robert, J.M. Marin, D.J. Balding, T. Guillemaud and A. Estoup. Inferring population history with DIYABC: a user-friendly approach to Approximate Bayesian Computations (2008) \emph{Bioinformatics}, \textbf{24} (23), 2713-2719.
\item  \textbf{version 1 :}  Cornuet J.M., V. Ravign\'e and A. Estoup, 2010. Inference on population history and model checking using DNA sequence and microsatellite data with the sofware DIYABC (v1.0) (2010) \emph{BMC Bioinformatics}, \textbf{11}, 401.
\item  \textbf{version 2 :}  Cornuet J.M., Pudlo P., Veyssier J., Dehne-Garcia A., V. Ravign\'e and A. Estoup. Improved inference on population history using SNP's and DIYABC (v2). \emph{submitted}
\end{itemize}
\subsection{Web site}
http://www.montpellier.inra.fr/CBGP/diyabc\\



\newpage
\input{methodology}
\input{gui_description}
\input{Implementation}
\input{biblio}
\end{document}